cmd/terminal.go

```diff
@@ -0,0 +1,367 @@
+	if err != nil {
+		return nil, err
+	}
+
+	skipVerify, err := cmd.Flags().GetBool(argRootSkipVerify)
+	if err != nil {
+		return nil, err
+	}
+
+	return &TerminalCmd{
+		server:     server,
+		skipVerify: skipVerify,
+		deviceID:   args[0],
+		stop:       make(chan bool),
+	}, nil
+}
+
+// Run executes the command
+func (c *TerminalCmd) Run() error {
+	ctx, cancelContext := context.WithCancel(context.Background())
+	defer cancelContext()
+
+	log.Info(fmt.Sprintf("Connecting to the remote terminal of the device %s...", c.deviceID))
+
+	tokenPath, err := getDefaultAuthTokenPath()
+	if err != nil {
+		return errors.Wrap(err, "Unable to determine the auth token path")
+	}
+
+	token, err := ioutil.ReadFile(tokenPath)
+	if err != nil {
+		return errors.Wrap(err, "Please Login first")
+	}
+
+	// get the terminal width and height
+	termID := int(os.Stdout.Fd())
+	termWidth, termHeight, err := terminal.GetSize(termID)
+	if err != nil {
+		return errors.Wrap(err, "Unable to get the terminal size")
+	}
+
+	// connect to the websocket
+	deviceConnectPath := "/api/management/v1/deviceconnect/devices/" + c.deviceID + "/connect"
+	parsedURL, err := url.Parse(c.server)
+	if err != nil {
+		return errors.Wrap(err, "Unable to parse the server URL")
+	}
+
+	scheme := getWebSocketScheme(parsedURL.Scheme)
+	u := url.URL{Scheme: scheme, Host: parsedURL.Host, Path: deviceConnectPath}
+
+	headers := http.Header{}
+	headers.Set("Authorization", "Bearer "+string(token))
+	websocket.DefaultDialer.TLSClientConfig = &tls.Config{
+		InsecureSkipVerify: c.skipVerify,
+	}
+	conn, _, err := websocket.DefaultDialer.Dial(u.String(), headers)
+	if err != nil {
+		return errors.Wrap(err, "Unable to connect to the device")
+	}
+	defer conn.Close()
+
+	// handle the ping-pong connection health check
+	err = conn.SetReadDeadline(time.Now().Add(pongWait))
+	if err != nil {
+		return errors.Wrap(err, "Unable to set the read deadline")
+	}
+
+	pingPeriod := (pongWait * 9) / 10
+	ticker := time.NewTicker(pingPeriod)
+	defer ticker.Stop()
+
+	conn.SetPongHandler(func(string) error {
+		ticker.Reset(pingPeriod)
+		return conn.SetReadDeadline(time.Now().Add(pongWait))
+	})
+
+	conn.SetPingHandler(func(msg string) error {
+		ticker.Reset(pingPeriod)
+		err := conn.SetReadDeadline(time.Now().Add(pongWait))
+		if err != nil {
+			return err
+		}
+		return conn.WriteControl(
+			websocket.PongMessage,
+			[]byte(msg),
+			time.Now().Add(writeWait),
+		)
+	})
+
+	log.Info("Press CTRL+] to quit the session")
+
+	// set the terminal in raw mode
+	oldState, err := term.MakeRaw(termID)
+	if err != nil {
+		return errors.Wrap(err, "Unable to set the terminal in raw mode")
+	}
+	defer func() {
+		_ = term.Restore(termID, oldState)
+	}()
+
+	// send the shell start message
+	m := &ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:   ws.ProtoTypeShell,
+			MsgType: wsshell.MessageTypeSpawnShell,
+			Properties: map[string]interface{}{
+				"terminal_width":  termWidth,
+				"terminal_height": termHeight,
```

Comment:
Maybe it's a good idea to use `textproto.CanonicalMIMEHeaderKey` to keep Properties keys consistent? Or should we commit to using snake_case?

Reply:
No strong opinion, but I have the impression snake_case is more common for APIs. At least all three of Github, Gitlab and Slack are using it.

Reply:
I agree, `stake_case` seems a better option to me.